File: day3/solution.py
from typing import List, Set

with open('input.txt', 'r') as file:
    lines = file.readlines()
    lines = [l.strip() for l in lines]

    lower_case_letters = 'abcdefghijklmnopqrstuvwxyz'
    letters = f" {lower_case_letters}{lower_case_letters.upper()}"


    def _common_char_in_sets(*s: Set[str]) -> str:
        (char,) = set.intersection(*s)
        return char


    def find_common_char_in_line_halves(line: str) -> str:
        mid = len(line) // 2
        return _common_char_in_sets(set(line[:mid]), set(line[mid:]))


    def solve_part1() -> int:
        return sum(map(letters.index, map(find_common_char_in_line_halves, lines)))


    def chunks() -> List[List[str]]:
        return [lines[i:i + 3] for i in range(0, len(lines), 3)]


    def find_common_char_in_chunk(chunk: List[str]) -> str:
        return _common_char_in_sets(*map(set, chunk))


    def solve_part2() -> int:
        return sum(map(letters.index, map(find_common_char_in_chunk, chunks())))


    print(solve_part1() == 7875)
    print(solve_part2() == 2479)

File: day3/solution.js
const fs = require('fs')
const {splitIntoChunks} = require("../helpers");

function processInput() {
    const rawData = fs.readFileSync('./day3/input.txt', 'utf8')
    return rawData.split('\n').map(line => line.split(''))
}

const lines = processInput()
const lower = 'abcdefghijklmnopqrstuvwxyz'
const allLetters = ' ' + lower + lower.toUpperCase()

function findCommonChar(line) {
    const firstHalf = line.slice(0, line.length / 2)
    const secondHalf = line.slice(line.length / 2)

    return firstHalf.find(char => secondHalf.includes(char))
}

function solvePart1() {
    return lines.reduce((acc, line) => {
        const commonChar = findCommonChar(line)
        return acc + allLetters.indexOf(commonChar)
    }, 0)
}

function findCommonCharPart2([line1, line2, line3]) {
    return line1.find(char => line2.includes(char) && line3.includes(char))
}

function solvePart2() {
    const chunks = splitIntoChunks(lines,3)

    return chunks.reduce((acc, chunk) => {
        const commonChar = findCommonCharPart2(chunk)
        return allLetters.indexOf(commonChar) + acc
    }, 0)
}

console.log(solvePart1() == 7875)
console.log(solvePart2() == 2479)